File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
def solution(files):
    # 방법 1.헤더, 넘버, 테일 구분한다음
    # 딕셔너리에 value로 정렬한다음 키를 반환하자
    # 방법 2. 바로 정렬 갈기고 반환하자
    file_dic={}
    while files:
        a = files.pop(0)
        head = ""
        temp = ""
        for i in a:
            if i.isdigit():
                temp +=i
            # temp 나올 때까지
            elif not temp: 
                head+=i
                pass
            else:
                break
        file_dic[a] = (head.upper(),int(temp))
    return sorted(file_dic.keys(), key = lambda x: (file_dic[x]))

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.js
function solution(files) {
    var answer = [];
    let list = []
    for(const file of files){
        let head = ''
        let number = ''
        for(const f of file){
            if(!isNaN(parseInt(f)) && number.length < 6){
                number += f
            }
            else{
                if(number){
                    break
                }
                head += f
            }
        }
        list.push([head.toLowerCase(), Number(number), file])
    }
    list.sort((a, b) => {
    if(a[0] !== b[0]){
        return a[0] > b[0] ? 1 : -1
    }
    else{
        if(a[1] !== b[1]){
            return a[1] > b[1] ? 1 : -1
        }
        else{
            return files.indexOf(a[2]) > files.indexOf(b[2]) ? 1 : -1
        }
    }
    })
    for(l of list){
        answer.push(l[2])
    }
    return answer;
}